### PR TITLE
Add `Scoped` `Plugin`

### PR DESCRIPTION
**Pull request by hlbarber**
## Motivation and Context

The [`FilterByOperationName`](https://docs.rs/aws-smithy-http-server/0.55.4/aws_smithy_http_server/plugin/struct.FilterByOperationName.html) allows the customer to filter application of a plugin. However this is a _runtime_ filter. A faster and type safe alternative would be a nice option.

## Description

Add `Scoped` `Plugin` and `scope` macro.

### CODE REVIEW

File: rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::marker::PhantomData;
+
+use super::Plugin;
+
+/// Marker struct for `true`.
+///
+/// Implements [`ConditionalApply`] which applies the a [`Plugin`].
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs:12]:
```suggestion
/// Implements [`ConditionalApply`] which applies the [`Plugin`].
```

File: rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    Pl: Plugin<P, Op, S>,
+{
+    type Service = <Scope::Contains as ConditionalApply<P, Op, S, Pl>>::Service;
+
+    fn apply(&self, svc: S) -> Self::Service {
+        <Scope::Contains as ConditionalApply<P, Op, S, Pl>>::apply(&self.plugin, svc)
+    }
+}
+
+/// A macro for creating subsets operations for [`Scoped`].
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs:104]:
```suggestion
/// A macro to help with scoping plugins to a subset of all operations.
```

Maybe?

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 3]
```diff
@@ -111,6 +111,8 @@ mod filter;
 mod identity;
 mod layer;
 mod pipeline;
+#[doc(hidden)]
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:114]: Why hidden?

File: rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+///
+/// The scope must include or exclude _all_ operations.
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs:106]:
```suggestion
/// The scope must partition _all_ operations, that is, each and every operation must be included or excluded, but not both.
```

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs (hunk 3 above)
Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:114]: I'd like `Scoped` and the `scope` macros to be the primary API and the `#[doc(hidden)]` items to be implementation details.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ScopeMacroGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    /** Calculate all `operationShape`s contained within the `ServiceShape`. */
+    private val index = TopDownIndex.of(codegenContext.model)
+    private val operations = index.getContainedOperations(codegenContext.serviceShape).toSortedSet(compareBy { it.id })
+
+    fun macro(): Writable = writable {
+        val firstOperationName = codegenContext.symbolProvider.toSymbol(operations.first()).name
+        val operationNames = operations
+            .map { codegenContext.symbolProvider.toSymbol(it).name }
+            .joinToString(" ")
```
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ScopeMacroGenerator.kt:33]:
```suggestion
                val operationNames = operations.joinToString(" ") { codegenContext.symbolProvider.toSymbol(it).name }
```

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ScopeMacroGenerator.kt [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            /// A macro to help with scoping plugins to a subset of all operations.
+            ///
+            /// In contrast to [`aws_smithy_http_server::scope`](#{SmithyHttpServer}::scope), this macro has knowledge
+            /// of the service and any operations _not_ specified will be placed in the opposing group.
+            ///
+            /// ## Example
+            ///
+            /// ```rust
+            /// scope! {
+            ///     /// Includes [`$firstOperationName`], excluding all other operations
```
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ScopeMacroGenerator.kt:60]:
```suggestion
            ///     /// Includes [`$firstOperationName`], excluding all other operations.
```

File: rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    Pl: Plugin<P, Op, S>,
+{
+    type Service = <Scope::Contains as ConditionalApply<P, Op, S, Pl>>::Service;
+
+    fn apply(&self, svc: S) -> Self::Service {
+        <Scope::Contains as ConditionalApply<P, Op, S, Pl>>::apply(&self.plugin, svc)
+    }
+}
+
+/// A macro to help with scoping plugins to a subset of all operations.
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/plugin/scoped.rs:103]: Link "plugins" to the `aws_smithy_http_server::plugin` module?

**Pull request merged**